spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+                    timeout=PARALLEL_GROUP_STATUS_INTERVAL * 20
+                )
+            except Exception:
+                # No more messages are being received: the whole group has hung
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name in ("starting", "running")
+                ):
+                    other_cmd_info.change_state("hung")
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name == "pending"
+                ):
+                    other_cmd_info.change_state("cancelled")
+                break
+            cmd_info = cmd_infos[cast(int, mess["cmd_index"])]
+            if mess["type"] in ("started", "keepalive"):
+                cmd_info.last_keepalive_time = int(time())
+            for other_cmd_info in (
+                c for c in cmd_infos if c.state.name in ("starting", "running")
+            ):
+                if (
+                    other_cmd_info.last_keepalive_time is not None
+                    and time() - other_cmd_info.last_keepalive_time
+                    > PARALLEL_GROUP_STATUS_INTERVAL * 20
+                ):
+                    # a specific command has hung
+                    other_cmd_info.change_state("hung")
+            if mess["type"] == "started":
+                cmd_info.change_state("running")
+                cmd_info.running_os_cmd_index = cast(int, mess["os_cmd_index"])
+                cmd_info.pid = cast(int, mess["pid"])
+            if mess["type"] == "completed":
+                cmd_info.rc = cast(int, mess["rc"])
+                if cmd_info.rc == 0:
+                    cmd_info.change_state("succeeded")
+                    if not dry:
+                        update_lockfile(current_dir, cmd_info.cmd)
+                    next_worker_process = next(worker_process_iterator, None)
+                    if next_worker_process is not None:
+                        _start_worker_process(next_worker_process, proc_to_cmd_infos)
+                elif cmd_info.rc > 0:
+                    cmd_info.change_state("failed")
+                else:
+                    cmd_info.change_state("terminated")
+                cmd_info.console_output = cast(str, mess["console_output"])
+            if any(
+                c for c in cmd_infos if c.state.name in ("failed", "terminated", "hung")
+            ):
+                # a command in the group hasn't succeeded, so terminate/cancel the rest
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name == "running"
+                ):
+                    try:
+                        os.kill(cast(int, other_cmd_info.pid), SIGTERM)
+                    except:
+                        # there could be a race condition between the process completing
+                        # and the message arriving to notify about it
+                        pass
```

Comment:
Could you elaborate on what happens with the stray process in this case?

Reply:
I'll update the comment in the code, but the list of processes to kill is based on the main process' knowledge of what the worker processes are doing. It could be the case that a subprocess has already completed but that the message from the worker process to the main process notifying about this is still in the queue.

Reply:
Thanks for the context and the updated comment.